### PR TITLE
Share Lease Feature

### PR DESCRIPTION
**Pull request by tasherif-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py [hunk 1]
```diff
@@ -266,7 +266,7 @@ def from_connection_string(
 
     @distributed_trace
     def acquire_lease(self, lease_id=None, **kwargs):
-        # type: (int, Optional[str], **Any) -> BlobLeaseClient
+        # type: (Optional[str], **Any) -> ShareLeaseClient
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py:269]: good catch

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py [hunk 2]
```diff
@@ -283,12 +283,12 @@ def acquire_lease(self, lease_id=None, **kwargs):
 
         .. admonition:: Example:
 
-            .. literalinclude:: ../samples/blob_samples_common.py
-                :start-after: [START acquire_lease_on_blob]
-                :end-before: [END acquire_lease_on_blob]
+            .. literalinclude:: ../samples/file_samples_client.py
+                :start-after: [START acquire_lease_on_file]
+                :end-before: [END acquire_lease_on_file]
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py:288]: Good catch! if this is changed the mark in file_samples_client.py should also be changed

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 3]
```diff
@@ -137,7 +201,15 @@ def change(self, proposed_lease_id, **kwargs):
...
                 timeout=kwargs.pop('timeout', None),
                 cls=return_response_headers,
+                **kwargs) if isinstance(self._client, FileOperations) \
+                else self._client.change_lease(
+                lease_id=self.id,
+                proposed_lease_id=proposed_lease_id,
+                timeout=kwargs.pop('timeout', None),
+                sharesnapshot=self.snapshot,
+                cls=return_response_headers,
                 **kwargs)
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:211]: you don't have to check if it's a FileOperations instance, since when you call acquire_lease on the given self._client, the corresponding acquire_lease will be auto detected and called no matter it's in FileOperations or ShareOperations

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 4]
```diff
@@ -110,7 +167,14 @@ def release(self, **kwargs):
                 lease_id=self.id,
                 timeout=kwargs.pop('timeout', None),
                 cls=return_response_headers,
+                **kwargs) if isinstance(self._client, FileOperations) \
+                else self._client.renew_lease(
+                lease_id=self.id,
+                timeout=kwargs.pop('timeout', None),
+                sharesnapshot=self.snapshot,
+                cls=return_response_headers,
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:175]: these checking conditions could be removed.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py [hunk 5]
```diff
@@ -366,6 +402,7 @@ def delete_share(
                 :dedent: 12
                 :caption: Deletes the share and any snapshots.
         """
+        access_conditions = get_access_conditions(kwargs.pop('lease', None))
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py:405]: missing doc for these apis

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/aio/_lease_async.py [hunk 6]
```diff
@@ -106,7 +158,14 @@ async def release(self, **kwargs):
                 lease_id=self.id,
                 timeout=kwargs.pop('timeout', None),
                 cls=return_response_headers,
+                **kwargs) if isinstance(self._client, FileOperations) \
+                else await self._client.renew_lease(
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/aio/_lease_async.py:162]: this is calling a wrong api

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 3 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:211]: Yes, I'm aware of that however the reason I'm checking if its a FileOperation or a ShareOperation is because the spec stated that only Shares can pass snapshots and files cannot

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/aio/_lease_async.py (hunk 6 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/aio/_lease_async.py:162]: big woops

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 3 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:211]: if you pass sharesnapshot=self.snapshot from ShareClient, then lease_client.acuqire_lease will have that snapshot info in kwargs.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py (hunk 5 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py:405]: You are absolutely right! I'll add that

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 3 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:211]: you are absolutely right, I completely forgot about this! thank you :)

Comment (xiafu-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py]: Another thing, we fileshare doesn't support write, I'm still thinking if it's necessary to acquire lease on a snapshot

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py [hunk 7]
```diff
@@ -353,8 +389,22 @@ def delete_share(
 
         :param bool delete_snapshots:
             Indicates if snapshots are to be deleted.
+        :keyword lease:
+            Required if the share has an active lease. Value can be a ShareLeaseClient object
+            or the lease ID as a string.
+
+            .. versionadded:: 12.2.0
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py:396]: not 12.2.0. it should be 12.6.0

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 3 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py]: BTW currently, the code doesn't work, since ShareClient isn't passing sharesnapshot=self.snapshot when calling lease_client.acquire_lease() internally, while it doesn't matter, as long as we detect it in LeaseClient.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 8]
```diff
@@ -125,6 +125,7 @@ def renew(self, **kwargs):
             eg. "\"tagname\"='my tag'"
 
             .. versionadded:: 12.4.0
+
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:128]: good catch

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 9]
```diff
@@ -90,22 +105,60 @@ def acquire(self, **kwargs):
...
+        the lease may be renewed even if it has expired as long as the share
+        has not been leased again since the expiration of that lease. When you
+        renew a lease, the lease duration clock resets.
+
+        :keyword str etag:
+            An ETag value, or the wildcard character (*). Used to check if the resource has changed,
+            and act according to the condition specified by the `match_condition` parameter.
+        :keyword ~azure.core.MatchConditions match_condition:
+            The match condition to use upon the etag.
+        :keyword str if_tags_match_condition
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:123]: check these parameters doc and remove the inaccurate ones.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 10]
```diff
@@ -90,22 +105,60 @@ def acquire(self, **kwargs):
...
+            Specify a SQL where clause on share tags to operate only on share with a matching value.
+            eg. "\"tagname\"='my tag'"
+
+            .. versionadded:: 12.4.0
 
+        :keyword int timeout:
+            The timeout parameter is expressed in seconds.
+        :return: None
+        """
+        if isinstance(self._client, FileOperations):
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:136]: good job on checking this

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 11]
```diff
@@ -146,8 +201,8 @@ def change(self, proposed_lease_id, **kwargs):
 
     @distributed_trace
     def break_lease(self, **kwargs):
-        # type: (Optional[int], Any) -> int
-        """Force breaks the lease if the file has an active lease. Any authorized request can break the lease;
+        # type: (Optional[int, str], Any) -> int
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:207]: what is str for?

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 9 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:126]: Yep, going through their docstrings at the moment since I learned about versioning in docstrings

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py [hunk 12]
```diff
@@ -353,8 +389,22 @@ def delete_share(
...
+        :paramtype lease: ~azure.storage.fileshare.ShareLeaseClient or str
         :keyword int timeout:
             The timeout parameter is expressed in seconds.
+        :keyword lease:
+            Required if the share has an active lease. Value can be a ShareLeaseClient object
+            or the lease ID as a string.
+
+            .. versionadded:: 12.6.0
+
+        :paramtype lease: ~azure.storage.fileshare.ShareLeaseClient or str
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_share_client.py]: duplicate

File: sdk/storage/azure-storage-file-share/tests/test_share.py [hunk 13]
```diff
@@ -184,6 +184,232 @@ def test_undelete_share(self, resource_group, location, storage_account, storage
...
+        # Act
+        props = share_client.get_share_properties(lease=lease_id)
+        lease_id.break_lease()
+
+        # Assert
+        self.assertIsNotNone(props)
+        self.assertDictEqual(props.metadata, metadata)
+        self.assertEqual(props.lease.duration, 'infinite')
+        self.assertEqual(props.lease.state, 'leased')
+        self.assertEqual(props.lease.status, 'locked')
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/tests/test_share.py:351]: Can we add a test for list blob like this to make sure the lease is parsed properly when calling from_generated()

File: sdk/storage/azure-storage-file-share/tests/test_share.py [hunk 14]
```diff
@@ -184,6 +184,232 @@ def test_undelete_share(self, resource_group, location, storage_account, storage
...
+    @GlobalStorageAccountPreparer()
+    def test_lease_share_twice(self, resource_group, location, storage_account, storage_account_key):
+        self._setup(storage_account, storage_account_key)
+        share_client = self._create_share('test')
+
+        # Act
+        lease = share_client.acquire_lease(lease_duration=15)
+
+        # Assert
+        lease2 = share_client.acquire_lease(lease_id=lease.id)
```
[the remark this answers is not in the document]
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/tests/test_share.py:272]: Yes, we can. As per my understanding as long as we provide the lease id then we're able to acquire it again

File: sdk/storage/azure-storage-file-share/tests/test_share.py (hunk 13 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/tests/test_share.py:351]: I'm not sure I follow, this is a test for share leases not blobs?

Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/tests/test_share.py:354]: I have added two more tests for listing shares and ensuring that lease is being parsed as a share property

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 15]
```diff
@@ -61,25 +68,33 @@ def __exit__(self, *args):
         self.release()
 
     @distributed_trace
-    def acquire(self, **kwargs):
+    def acquire(self, lease_duration=-1, **kwargs):
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:71]: I'm good with this! Since it's not a breaking change. While I still feel it would be safer to ask for Anna's opinion @annatisch are you good with this?

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py [hunk 16]
```diff
@@ -283,15 +283,15 @@ def acquire_lease(self, lease_id=None, **kwargs):
...
+                :end-before: [END acquire_and_release_lease_on_file]
                 :language: python
-                :dedent: 8
-                :caption: Acquiring a lease on a blob.
+                :dedent: 12
+                :caption: Acquiring a lease on a file.
         """
         lease = ShareLeaseClient(self, lease_id=lease_id)  # type: ignore
-        lease.acquire(**kwargs)
+        lease.acquire(lease_duration=-1, **kwargs)
```
Comment (annatisch) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py:294]: Do we not support a user passing in this value?

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 15 above)
Comment (annatisch) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:71]: Could this not just be rolled into kwargs?
I don't really see a reason it needs to be in method signature...

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 17]
```diff
@@ -123,15 +167,16 @@ def change(self, proposed_lease_id, **kwargs):
         """ Changes the lease ID of an active lease. A change must include the current lease ID in x-ms-lease-id and
         a new lease ID in x-ms-proposed-lease-id.
 
-
         :param str proposed_lease_id:
-            Proposed lease ID, in a GUID string format. The File service returns 400
+            Proposed lease ID, in a GUID string format. The File or Share service returns 400
```
Comment (annatisch) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:171]: Maybe rephrase this to say something like "An error will be raised if the proposed lease ID is not in the correct format".
The actual status code is probably not important to a customer.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py [hunk 18]
```diff
@@ -145,22 +190,42 @@ def change(self, proposed_lease_id, **kwargs):
         self.last_modified = response.get('last_modified')   # type: datetime
 
     @distributed_trace
-    def break_lease(self, **kwargs):
+    def break_lease(self, lease_break_period=None, **kwargs):
```
Comment (annatisch) [inline sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:193]: This parameter should also be moved into kwargs - especially as it only applies to Shares

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py (hunk 16 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_file_client.py:294]: We do not support it for files. The file lease duration can only be infinite while in share leases it can be set

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 15 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:71]: This would be inconsistent to `\blob\_lease.py`, do we not want to maintain the same method definition/pattern as `\blob\_lease.py`?

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 18 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:193]: > This would be inconsistent to `\blob\_lease.py`, do we not want to maintain the same method definition/pattern as `\blob\_lease.py`?

this comment applies here as well

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 17 above)
Comment (tasherif-msft) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:171]: I agree, sounds good :) will add that

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 18 above)
Comment (annatisch) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:193]: In this case it should be in kwargs.
It's still consistent with blobs in that the parameter is named the same - so from a user perspective it's consistent.

File: sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py (hunk 15 above)
Comment (annatisch) [reply sdk/storage/azure-storage-file-share/azure/storage/fileshare/_lease.py:71]: It's not really inconsistent with Blobs, because the parameter is the same. But if this parameter only applies to Share leases, we should not make it so prominent in the method signature. I think kwargs would be best :)

**Review by annatisch — APPROVED** [review]
LGTM
